BUG: param_parse couldn't handle nested parentheses

eia-akahan opened the pull request:
BUG:  param_parse couldn't handle nested parentheses - see https://github.com/numpy/numpy/issues/28095.  

This adds a function that determines the appropriate order of nested parentheses and  properly assigns the dimensions based on the order of operations. It solves the case outlined in https://github.com/numpy/numpy/issues/28095, so that all parameters are mapped properly and assigned values.

BUG: param_eval expects v to be a string, but get_parameters sets v to integers in 2 cases.

```
            if real16pattern.search(v):
                v = 8
            elif real8pattern.search(v):
                v = 4
```

by enclosing those values in single quotes, now v is a string, which removes the error message: "param_eval: got eval() arg 1 must be a string, bytes or code object on 4." which is triggered when those two if statements are activated.

This is my first pull request for numpy - I'm not sure if I need to create test cases.


<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

seberg commented on numpy/f2py/crackfortran.py:
```diff
@@ -3022,6 +3022,52 @@ def param_eval(v, g_params, params, dimspec=None):
+    string_equation=re.sub(r'\b[a-zA-Z_][a-zA-Z0-9_]*\b',
+                  lambda match: str(params[match.group(0)])
+                                if match.group(0) in params
+                                else _raise_value_error(match.group(0)),
```
I would just make inline `def ...` or move all into the helper. Now we have a helper but just to raise an error.

seberg commented on numpy/f2py/crackfortran.py:
```diff
@@ -3082,12 +3128,9 @@ def param_parse(d, params):
     if "(" in d:
         # this dimension expression is an array
-        dname = d[:d.find("(")]
-        ddims = d[d.find("(") + 1:d.rfind(")")]
```
Do you know what you are getting here, i.e. what `d` is/can be?  Since the code here works, I would expect that this is pretty well formatted and the balancing might not be accounting for much here?

I think I am getting at the parentheses balancing seems a bit odd and I would like to know if there is even a reason for it.

(Also nit, but `replace_variables_in_equation` isn't recursive :).)

seberg reviewed (commented):
Some comments to get things started, I hope someone more famliar with f2py will chime in.
About adding a test: Yes please do unless it is incredibly painful, but I suspect it is easy (maybe even just modifying/extending an existing test a bit).

HaoZeke reviewed (changes requested):
Thanks for working on this @eia-akahan. Sorry for the late response. The core bug (`param_parse` treats `(a + b)` as array indexing) is correctly identified.

However the current approach replaces working logic with a more complex engine that has some issues:

1. `balance_parentheses` should not be needed. The dimension pieces from `markoutercomma` are already balanced. Silently "fixing" unbalanced parens masks upstream bugs.

2. The `except TypeError: int_dim ` int(balanced_dim.split('(')[0])` fallback is fragile. For expressions starting with `(` this would do `int("")` and crash. It works for `myparamarray(nested(dim))` by coincidence.

3. The `{ in str(params[...])` heuristic for detecting array parameters couples to Python dict repr formatting.

A simpler fix: check `dname` (text before the first `(`). 
- When `dname` is empty the parens are grouping, not array indexing, so strip outer parens and recurse.
- When `dname` is non-empty, keep the existing array-index logic.

The `v = '8'` / `v = '4'` fix for `real16pattern` / `real8pattern` is correct and could go as a separate commit.

Here is a concrete alternative (happy to co-author / cleanup and finalize this):

```python
if "(" in d:
    dname ` d[:d.find("(")]
    if dname:
        # Array parameter indexing, e.g. pa(1)
        ddims ` d[d.find("(") + 1:d.rfind(")")]
        index ` int(param_parse(ddims, params))
        return str(params[dname][index])
    else:
        # Grouping parentheses, e.g. (a + b)
        d ` d[d.find("(") + 1:d.rfind(")")]
        return param_parse(d, params)
```

This preserves all existing behavior for array parameters and adds the missing case. The `v = '8'` fix should be a separate commit.

HaoZeke commented:
@eia-akahan sorry for the delay here, do you mind if I make a few changes to finish this up?

eia-akahan commented:
> @eia-akahan sorry for the delay here, do you mind if I make a few changes to finish this up?

@HaoZeke , That would be great.  Thanks for the review.

HaoZeke committed:
BUG: f2py: distinguish grouping parens from array index in param_parse

`param_parse` treats any dimension token containing "(" as array-parameter
indexing like `pa(1)`, computing `int(param_parse(inner))` to index
`params[dname]`. Grouping parentheses such as
`(mx_supply_curves + mx_intl_curves)` have an empty name before "("; the
inner expression then becomes a non-integer string and `int(...)` raises.
The exception is swallowed in `analyzevars`, leaving parameters
unsubstituted (regression since 1.26.4; gh-28095).

Branch on the text before the first "(": non-empty name keeps the
array-index path; empty name falls through to the general textual
parameter-substitution path, which preserves compound expressions like
`(n + 1)*m`.

Adds pure crackfortran tests and a minimal Fortran source fixture.

Closes #28095.
Supersedes #29857.

Co-authored-by: Ari Kahan <ari.kahan@eia.gov>
